### ISSUE TITLE
`mx.metric.Perplexity(-1) ` update_metric raise segement fault

### ISSUE THREAD

**Issue by saicoco** [dmlc/mxnet#5250]
For bugs or installation issues, please provide the following information.
The more information you provide, the more likely people will be able to help you.

## Environment info
Operating System:
ubuntu15.10
Compiler:
gcc-9.4.x
Package used (Python/R/Scala/Julia):
Python
MXNet version:
0.9.4
Or if installed from source:
yeap
MXNet commit hash (`git rev-parse HEAD`):
be38c5b84030a63d0ab51f19737f99a75a7feb23

If you are using python package, please provide
anaconda4.2
Python version and distribution:

If you are using R package, please provide

R `sessionInfo()`:

## Error Message:
Please paste the full error message, including stack trace.
```
Segmentation fault (core dumped)

[Done] exited with code=139 in 39.916 seconds
```
## Minimum reproducible example
if you are using your own code, please provide a short script that reproduces the error.
```python
def caption_module(num_lstm_layer, seq_len, vocab_size, num_hidden, num_embed, batch_size, dropout=0.):

    seq = mx.sym.Variable('word_data')
    label = mx.sym.Variable('softmax_label')
    image_feature = mx.sym.Variable('image_feature')
    image_embed = mx.sym.FullyConnected(data=image_feature, num_hidden=num_embed, name='img_embed')
    word_embed = mx.sym.Embedding(data=seq, input_dim=vocab_size, output_dim=num_embed, name='seq_embed')

    # Concat image_embed with word_embed in axis=1 (batch_size, length+1, num_hidden)
    image_embed_reshape = mx.sym.expand_dims(image_embed, axis=1, name='image_embed_expand_dims')
    embedd_feature = mx.sym.Concat(image_embed_reshape, word_embed, num_args=2, dim=1, name='embed_concat')

    stack = mx.rnn.SequentialRNNCell()
    for i in range(num_lstm_layer):
        stack.add(mx.rnn.LSTMCell(num_hidden=num_hidden, prefix='lstm_l%d_' % i))
    outputs, states = stack.unroll(length=seq_len+1, inputs=embedd_feature, merge_outputs=True)

    outputs_crop = mx.sym.crop(outputs, begin=(0, 1, 0), end=(batch_size, seq_len+1, num_embed))
    pred = mx.sym.Reshape(outputs_crop, shape=(-1, num_hidden))
    pred = mx.sym.FullyConnected(pred, num_hidden=vocab_size, name='pred_fc')
    label = mx.sym.Reshape(label, shape=(-1,))
    softmax_output = mx.sym.SoftmaxOutput(data=pred, label=label, name='softmax')
    return softmax_output
```

Also, I custom my dataprovider, here `image_feature=(batch_size, 4096)`, `word_length=(batch_size, 37)`, `len(vocab)=8976`, my word are padding to 37, so I dont use BucketingModule.

## Steps to reproduce
```python
lstm = caption_module(num_lstm_layer=num_lstm_layer, seq_len=train_data.sent_length,
                          vocab_size=train_data.vocab_size, num_hidden=num_hidden, num_embed=num_embed, batch_size=batch_size)

    caption = mx.mod.Module(symbol=lstm, data_names=['image_feature', 'word_data'], label_names=('softmax_label',), context=ctx)

    caption.fit(
        train_data=train_wrap,
        eval_data=val_wrap,
        eval_metric=mx.metric.Perplexity(-1),
        kvstore='device',
        optimizer='sgd',
        optimizer_params={'learning_rate': 0.01,
                          'momentum': 0.9,
                          'wd': 0.00001},
        initializer=mx.init.Xavier(factor_type="in", magnitude=2.34),
        num_epoch=20,
        batch_end_callback=mx.callback.Speedometer(batch_size, 1))
```
For above code, when it run into `fit` and run at the line where `update_metric(metric)`, it will raise segfault, when I comment it and it works well. My dataIter works well and I test it some times. So I think the error must be related to update_metric... However, the example--lstm_bucketing.py runs well..I dont what I should do .
